sdks/python/apache_beam/io/gcp/bigquery.py

```diff
@@ -659,27 +659,41 @@ def __init__(self, source, test_bigquery_client=None, use_legacy_sql=True,
     else:
       self.query = self.source.query
 
-  def _get_source_table_location(self):
-    tr = self.source.table_reference
-    if tr is None:
-      # TODO: implement location retrieval for query sources
-      return
+  def _get_source_location(self):
+    """
+    Get the source location (e.g. ``"EU"`` or ``"US"``) from either
 
-    if tr.projectId is None:
-      source_project_id = self.executing_project
-    else:
-      source_project_id = tr.projectId
+    - :data:`source.table_reference`
+      or
+    - The first referenced table in :data:`source.query`
+
+    See Also:
+      - :meth:`BigQueryWrapper.get_query_location`
+      - :meth:`BigQueryWrapper.get_table_location`
 
-    source_dataset_id = tr.datasetId
-    source_table_id = tr.tableId
-    source_location = self.client.get_table_location(
-        source_project_id, source_dataset_id, source_table_id)
-    return source_location
+    Returns:
+      Optional[str]: The source location, if any.
+    """
+    if self.source.table_reference is not None:
+      tr = self.source.table_reference
+      return self.client.get_table_location(
+          tr.projectId if tr.projectId is not None else self.executing_project,
+          tr.datasetId, tr.tableId)
+    elif self.source.query is not None:
+      return self.client.get_query_location(
+          self.executing_project,
+          self.source.query,
+          self.source.use_legacy_sql)
+    else:
+      logging.error(
+          "Unable to get source location for %r as it has neither "
+          "table nor query.", self.source)
+      return None
```

Comment:
Pushed a `raise` in https://github.com/apache/beam/pull/6463/commits/fb034a17897cdf1133df4ac568186d274b689c0f

Reply:
Perhaps this check is better placed in `BigQueryReader.__init__`, since it already makes the same assumption that either a table reference or query is set:
https://github.com/apache/beam/blob/5aacd6bcbf78d0b70ff3d289ad7fbc0a156a54c8/sdks/python/apache_beam/io/gcp/bigquery.py#L649-L660